### PR TITLE
Add an optional system property to set the version returned from main response to 7.10.2

### PR DESCRIPTION
**Pull request by mch2**
Signed-off-by: Marc Handalian <handalm@amazon.com>

### Description
Add an optional system property, opensearch.http.override_main_response_version,  that spoofs the version.number returned from MainResponse to 7.10.2 instead of the opensearch build version.
 
```
Test steps:
1. Start with param set:
OPENSEARCH_JAVA_OPTS="$OPENSEARCH_JAVA_OPTS -Dopensearch.http.override_main_response_version=true" ./bin/opensearch
2. curl -X GET "localhost:9200/?pretty"

{
  "name" : "test",
  "cluster_name" : "opensearch",
  "cluster_uuid" : "CHsbWvhASVW_uCQVnpC06g",
  "version" : {
    "distribution" : "opensearch",
    "number" : "7.10.2",
    "build_type" : "tar",
    "build_hash" : "9f54032a505381591b57375cd7c906496b977a97",
    "build_date" : "2021-06-04T19:48:23.702201Z",
    "build_snapshot" : true,
    "lucene_version" : "8.8.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  }
}

```

### Issues Resolved
[693]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
This works fine. 

Thinking out loud, I don't love that the version number is resolved in two different places, one by calling `build.getQualifiedVersion()` in `MainResponse#getVersionNumber` and the other in `TransportMainAction` constructor by calling `this.clientCompatibilityVersion = HttpTransportSettings.SETTING_HTTP_MAIN_RESPONSE_VERSION_NUMBER.get(settings);`. It might be cleaner to either pass `settings` into `TransportMainAction`, or if possibly maybe even default `SETTING_HTTP_MAIN_RESPONSE_VERSION_NUMBER` to `build.getQualifiedVersion()`. It's NBD either way.

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (dblock) [inline server/src/main/java/org/opensearch/action/main/MainResponse.java:147]: Alternatively, the need for this `if` tells me that it should have happened earlier our outside of here, aka either the input to `MainRespose` was the version we wanted to return, or `build.getVersionNumber()` should become `build.getResponseVersionNumber()` where all the logic happens. It's easier to unit test when the logic for the version number defaulting is self-contained.

File: server/src/test/java/org/opensearch/action/main/MainResponseTests.java
Comment (dblock) [inline server/src/test/java/org/opensearch/action/main/MainResponseTests.java:118]: This test should only be focused on the field, otherwise it's hard to tell what we're after and it can break for unrelated changes.

**Comment by dlvenable** [comment]
Thanks you, @mch2 . I activated this via:

```
./gradlew run -Dtests.opensearch.http.main.response.version=7.10.2
```

I was able to get Logstash 9.4 and 10.8 to save the mapping and the documents to OpenSearch with this change. Thus this change resolves #667 from my testing. I'm not the originating author of that, but I was able to reproduce that issue.

I also ran Filebeats 7.12 with that modification and it was able to create the mappings and write documents. No ERROR lines. Thus, it should resolve #706. Again, I'm not the author of that one either.

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (dlvenable) [reply server/src/main/java/org/opensearch/action/main/MainResponse.java:147]: It seems the `Build` is concerned with the build information and probably should not be the place. But, perhaps moving this condition into `TransportMainAction` would make sense.

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/main/TransportMainAction.java:57]: I still suggest using system property instead of node property in yml as this is an advance control and generally not recommended. It is a hack to unblock clients. Deprecating system property would be lot easier as compared to setting in yml and both will serve the same purpose.

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/main/TransportMainAction.java:61]: It should just be 7.10.2. what is the use case to give option to choose any version < 7.10.2. What if someone set 6.8. The code base will not automatically switch to 6.8 logic, this will add more confusion.

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (mch2) [reply server/src/main/java/org/opensearch/action/main/MainResponse.java]: Agree, I have removed the need for this by making the setting the default to the current build version.

File: server/src/test/java/org/opensearch/action/main/MainResponseTests.java
Comment (mch2) [reply server/src/test/java/org/opensearch/action/main/MainResponseTests.java]: updated this additional test to only check for the version number.

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java
Comment (mch2) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java:61]: Hardcoding this to 7.10.2 makes the assumption that the client is upgrading from 7.10.2.  We want to allow them to upgrade without forcing an upgrade to their client.

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java
Comment (mch2) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java:57]: agree - will update this

File: server/src/main/java/org/opensearch/action/main/TransportMainAction.java
Comment (dblock) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java:61]: What happens if I set the version to 6.8.0 and some client now believes I have a 6.x REST API? What was removed between 6 and 7?

Comment (mch2) [reply server/src/main/java/org/opensearch/action/main/TransportMainAction.java:61]: Good point, this is too much of a risk.  I don't think the original intent here of the setting is realistic and this should instead be a boolean to set to 7.10.2 directly similar to what is happening [here](https://github.com/opensearch-project/OpenSearch/blob/c62884a1e91606c3ddf663bf3b054148c6c493ec/server/src/main/java/org/opensearch/action/main/MainResponse.java#L106).

**Comment by mch2** [comment]
Updated this to be a system property and set directly in MainResponse.

@dlvenable would you mind validating this latest version?

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (dblock) [inline qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:1]: The name if this file and class is 💀 🔥

**Review by dblock — CHANGES_REQUESTED** [review]
I don't think you meant to call this `EvilSystemPropertyTests`, or did you?

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (mch2) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:1]: It is intentional, but I can't claim it as my own suggestion :)  There are existing similar tests under [/cluster/metadata](https://github.com/mch2/OpenSearch/blob/main/qa/evil-tests/src/test/java/org/opensearch/cluster/metadata/EvilSystemPropertyTests.java) and [cluster/routing](https://github.com/mch2/OpenSearch/blob/main/qa/evil-tests/src/test/java/org/opensearch/cluster/routing/EvilSystemPropertyTests.java)

Comment (dblock) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:1]: Oh wow, it's a thing!

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (dblock) [inline qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:34]: Needs to be in a try/finally block or a test failure will pollute following tests.

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (dblock) [inline server/src/main/java/org/opensearch/action/main/MainResponse.java:62]: `OVERRIDE_MAIN_RESPONSE_VERSION` to match `override_main_response_version`

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (shwetathareja) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:34]: you can also do the clean up in @After

Comment (dblock) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:34]: Except that runs after all the tests, so if we add more tests it will break the subsequent ones in the same class.

I don't envy whomever has the username `@after` on GitHub :)

Comment (mch2) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java]: updated with try/finally

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (dblock) [inline qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java:31]: This should wrap the `MainResponse` declaration above. It can fail as well.

**Review by dblock — APPROVED** [review]
Approved with nit.

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (shwetathareja) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java]: btw "After" runs after every test. "AfterClass" is run after all tests.

**Review by shwetathareja — APPROVED** [review]

**Review by shdubsh — APPROVED** [review]
LGTM!

File: qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java
Comment (dblock) [reply qa/evil-tests/src/test/java/org/opensearch/action/main/EvilSystemPropertyTests.java]: You're right of course. I personally dislike doing a certain cleanup in `after` unless it was setup in `before`. Just to keep things symmetrical. In Ruby we have things like `context` in which you can do "when the value of this setting is X", setup in before, then cleanup after. Not sure it's possible here.

**Review by dblock — APPROVED** [review]

**Comment by CEHENKLE** [comment]
start gradle check

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (nknize) [inline server/src/main/java/org/opensearch/action/main/MainResponse.java:63]:
```suggestion
        Booleans.parseBoolean(System.getProperty(OVERRIDE_MAIN_RESPONSE_VERSION_KEY), false);
```

File: server/src/main/java/org/opensearch/action/main/MainResponse.java
Comment (nknize) [inline server/src/main/java/org/opensearch/action/main/MainResponse.java:132]:
```suggestion
        if (OVERRIDE_MAIN_RESPONSE_VERSION) {
            DEPRECATION_LOGGER.deprecate(OVERRIDE_MAIN_RESPONSE_VERSION_KEY, "overriding version number will be removed " +
                "in version 2.0.0");
            builder.field("number", LegacyESVersion.V_7_10_2);
        } else {
            builder.field("number", build.getQualifiedVersion());
        }
```

**Review by nknize — CHANGES_REQUESTED** [review]
Let's add deprecation logging to this so user's know this isn't a permanent change.

I think this should be a cluster-wide setting that can be updated using `PUT /_cluster/settings`. Maybe make it a new DeprecatedResponse setting in `ClusterSettings`?  That default's to false and can be dynamically set using the cluster setting API?

**Comment by mch2** [comment]
Agree with deprecation logging, thanks for the callout.

As for making this a cluster setting, I originally cut this as an updatable cluster setting that defaults from a value in yml.  I then swapped it to only a sys prop because it was intended as an undocumented hack for a subset of users. It sounds like this has become larger and is going to live until 2.x so I agree with making rolling upgrades less error prone and reverting back to that (without requiring a default in yml) and give users more flexibility to toggle it on/off.

I'll hold this for about a day and if no strong objections I'll swap this to a cluster setting.

**Comment by dlvenable** [comment]
I imagine that the workflow for an OpenSearch user would look something like this:

1. Setup OpenSearch
2. Set the compatibility flag
3. Use OpenSearch with existing tools
4. At some point in the future, change tools plugins to OpenSearch versions; or updated versions which support both OpenSearch and ElasticSearch
5. Turn off compatibility
6. Use OpenSearch with working tools

So, I do think that a cluster setting which can be dynamically changed is the correct approach here.

Perhaps this would also help with updating an AES cluster.

**Comment by dblock** [comment]
Remote clusters were brought up in the community meeting. Let's address how this cluster-wide setting will support remote clusters as well.

**Comment by verbecee** [comment]
To @dblock 's point, we have a remote cluster architecture, where a single coordinating cluster connects to three standalone remote clusters (they don't know about each other/there is no cross-cluster replication). My thoughts were that the clusters would inherit their own version number through their own elasticsearch.yml/opensearch.yml file. Here is the cluster settings we use to connect the clusters. I don't recall seeing anything in the API documentation to pass along the version #.
```
{
  "persistent" : {
    "cluster" : {
      "remote" : {
        "cluster_zone_b" : {
          "mode" : "sniff",
          "skip_unavailable" : "true",
          "transport" : {
            "ping_schedule" : "30s",
            "compress" : "true"
          },
          "seeds" : [
<REDACTED>,
<REDACTED>,
<REDACTED>
          ]
        },
        "cluster_zone_c" : {
          "mode" : "sniff",
          "skip_unavailable" : "true",
          "transport" : {
            "ping_schedule" : "30s",
            "compress" : "true"
          },
          "seeds" : [
<REDACTED>,
<REDACTED>,
<REDACTED>
          ]
        },
        "cluster_zone_a" : {
          "mode" : "sniff",
          "skip_unavailable" : "true",
          "transport" : {
            "ping_schedule" : "30s",
            "compress" : "true"
          },
          "seeds" : [
<REDACTED>,
<REDACTED>,
<REDACTED>
          ]
        }
      },
      "routing" : {
        "rebalance" : {
          "enable" : "all"
        },
        "allocation" : {
          "allow_rebalance" : "always",
          "cluster_concurrent_rebalance" : "10",
          "node_concurrent_recoveries" : "12",
          "disk" : {
            "include_relocations" : "true",
            "threshold_enabled" : "true",
            "watermark" : {
              "low" : "90%",
              "flood_stage" : "96%",
              "high" : "95%"
            }
          },
          "exclude" : {
            "_ip" : "<REDACTED>"
          },
          "enable" : "all"
        }
      },
      "max_shards_per_node" : "3000"
    },
    "indices" : {
      "recovery" : {
        "max_bytes_per_sec" : "5gb"
      }
    }
  },
  "transient" : {
    "logger" : {
      "com" : {
        "amazon" : {
          "opendistroforelasticsearch" : {
            "security" : "DEBUG"
          }
        }
      }
    }
  }
}
```

**Comment by nknize** [comment]
> My thoughts were that the clusters would inherit their own version number through their own elasticsearch.yml/opensearch.yml file

How does `PUT /_cluster/settings` on each of the three clusters not solve this (instead of manually updating their .yml and doing a cluster restart)?

**Comment by verbecee** [comment]
> > My thoughts were that the clusters would inherit their own version number through their own elasticsearch.yml/opensearch.yml file
> 
> How does `PUT /_cluster/settings` on each of the three clusters not solve this (instead of manually updating their .yml and doing a cluster restart)?

That's fine too. Either way, I am trying to say that a multi-cluster architecture would not have any issue with this explicit versioning. dblock asked me to put our own use case/architecture after it was brought up on the community meeting

**Comment by mch2** [comment]
I have opened a new PR #847 to implement this as a cluster setting.  @verbecee This will actually allow you to set a default via opensearch.yml or set it with the cluster settings API.

Going to close this PR in preference of #847

**Pull request closed without merging**
